DOCS: New Guide: Upstream mTLS (#2814)
pomerium/pomerium: docs/guides/upstream-mtls.md

---
title: Upstream mTLS
lang: en-US
meta:
  - name: keywords
    content: pomerium identity-access-proxy mtls client-certificate
description: >-
  This guide covers how to configure Pomerium to provide mutual authentication
  (mTLS) to an upstream service, using client certificates with a custom certificate authority.
---

# Upstream mTLS With Pomerium

> Part of a complete zero-trust security model is secure communication between your authenticated proxy (Pomerium) and the upstream service it provides access to. This means both Pomerium *and* the upstream service will authenticate each other.

Reviewer: ```suggestion
Part of a complete zero-trust security model is secure communication between your identity-aware access proxy (Pomerium) and the upstream service it provides access to. This means both Pomerium *and* the upstream service will authenticate each other.
```

> Pomerium confirms the identity of the upstream service by the TLS certificate it serves. See [`tls_custom_ca_file`] and [`tls_server_name`] for more information on configuring Pomerium to accept an upstream's TLS certificate.

Reviewer: ```suggestion
Pomerium confirms the identity of an upstream service by the TLS certificate it serves. See [`tls_custom_ca_file`] and [`tls_server_name`] for more information on configuring Pomerium to accept an upstream's TLS certificate.
```

> How an upstream service is configured to validate and accept client certificates is entirely up to that specific service and how it's configured. To provide a general explanation, in this guide we will use [OpenSSL] and [mkcert] to demonstrate how to configure Pomerium to provide the client certificate.

Reviewer: ```suggestion
Generally, every organization has a preferred method of managing internal certificates. How an upstream service is configured to validate and accept client certificates is entirely up to that specific service and how it's configured. To provide a general explanation, in this guide we will use [OpenSSL] and [mkcert] to demonstrate how to configure Pomerium to provide the client certificate.
```

Reviewer: I think what we want to say here is:

- Managing internal TLS tends to be complicated (because you can't just use something like lets encrypt)
- Pomerium is flexibly enough to use whatever method of internal TLS the organization uses; whether that's using a service mesh, manual, or some sort of automated certificate management system

> How an upstream service is configured to validate and accept client certificates is entirely up to that specific service and how it's configured. To provide a general explanation, in this guide we will use [OpenSSL] and [mkcert] to demonstrate how to configure Pomerium to provide the client certificate.

Author: I'm not 100% on this change. We're not just using OpenSSL to provide the certificates. OpenSSL is acting as the upstream service itself, responding to traffic from Pomerium. How about:

```suggestion
How an upstream service is configured to validate and accept client certificates is entirely up to that specific service and how it's configured. To provide a general explanation, in this guide we will use [mkcert] to provide the client certificate to Pomerium, and [OpenSSL] as the upstream service.
```

> How an upstream service is configured to validate and accept client certificates is entirely up to that specific service and how it's configured. To provide a general explanation, in this guide we will use [OpenSSL] and [mkcert] to demonstrate how to configure Pomerium to provide the client certificate.

Reviewer: Agreed

> How an upstream service is configured to validate and accept client certificates is entirely up to that specific service and how it's configured. To provide a general explanation, in this guide we will use [OpenSSL] and [mkcert] to demonstrate how to configure Pomerium to provide the client certificate.

Author: Resolved in a new suggestion, as this one was made stale by the other change to this line.

> Generally, every organization has a preferred method of managing internal certificates. How an upstream service is configured to validate and accept client certificates is entirely up to that specific service and how it's configured. To provide a general explanation, in this guide we will use [OpenSSL] and [mkcert] to demonstrate how to configure Pomerium to provide the client certificate.

Author: Recreating the last suggestion from a thread on an outdated version:
```suggestion
Generally, every organization has a preferred method of managing internal certificates. How an upstream service is configured to validate and accept client certificates is entirely up to that specific service and how it's configured. To provide a general explanation, in this guide we will use [mkcert] to provide the client certificate to Pomerium, and [OpenSSL] as the upstream service.
```

## Before You Begin

> - To simplify this proof-of-concept demonstration, we assume that Pomerium is running in all-in-one mode as a system service with the configuration file in the [standard location](/docs/install/binary.md#configuration-file), configured to communicate with an identity provider (**IdP**), and all command are run on the same host. You may have to move files and/or adjust paths and commands to match a different configuration.

Author: Linked in 8c3f26f0004c2ebc38c2b63524388305e21349d9

    See [Install - Binaries](/docs/install/binary.md) for more information on initial setup.

- This guide uses [mkcert] to provide certificates. mkcert is a development tool, and not designed for production environments. In this guide, mkcert takes the place of your trusted certificate tooling solution.

- We use `*.pomerium.localhost.io` as the placeholder for the Pomerium domain space. If you're testing on your local computer you can edit `/etc/hosts` or use dnsmasq on [macOS](https://blog.thesparktree.com/local-development-with-wildcard-dns) or [Linux](https://sixfeetup.com/blog/local-development-with-wildcard-dns-on-linux) systems to route this domain space to your local computer.

## Create Certificates

!!!include(install-mkcert.md)!!!

> 1. Create a certificate and key for our demo upstream service, OpenSSL:

Reviewer: ```suggestion
1. Create a certificate and key for our example upstream service, OpenSSL:
```

    ```bash{1}
    mkcert openssl.localhost

    Created a new certificate valid for the following names 📜
    - "openssl.localhost"

    The certificate is at "./openssl.localhost.pem" and the key at "./openssl.localhost-key.pem" ✅

    It will expire on 10 March 2024 🗓
    ```

1. Create a client certificate and key for Pomerium to use:

    ```bash{1}
    mkcert -client 'pomerium@localhost'

    Created a new certificate valid for the following names 📜
    - "pomerium@localhost"

    The certificate is at "./pomerium@localhost-client.pem" and the key at "./pomerium@localhost-client-key.pem" ✅

    It will expire on 10 March 2024 🗓
    ```

> 1. Change ownership of the client certificate files and move them to Pomerium's configuration directory:
> 
>     ```bash
>     sudo chown pomerium:pomerium pomerium@localhost-client*pem /etc/pomerium
>     sudo mv pomerium@localhost-client*pem /etc/pomerium/
>     ```

Reviewer 2: I think we can skip the move step since file references can be relative.  Just use `./` instead of `/etc/pomerium`.  

```suggestion
```

Author: Per discussion, this is required for the assumed environment. Clarified by ed424ac6e4c0fe5aac926fafde4685e59fc915bb

1. In a terminal environment and the same directory where you created the certificate files, start an OpenSSL server process. Note that it will run in the foreground until stopped, so you will need another terminal environment to run additional commands:

    ```bash
    openssl s_server -key ./openssl.localhost-key.pem -cert ./openssl.localhost.pem -accept 44330 -www
    ```

    You can confirm that the server is responding using `curl`:

    ```
    curl -k https://localhost:44330
    <HTML><BODY BGCOLOR="#ffffff">
    <pre>

    s_server -key ./openssl.localhost-key.pem -cert ./openssl.localhost.pem -accept 44330 -www
    Secure Renegotiation IS supported
    Ciphers supported in s_server binary
    ...
    ```

    ::: tip
    Even though OpenSSL is providing a certificate signed by a CA in our trust store, we need the `-k` flag because the domain name `localhost` doesn't match `openssl.localhost` defined in our certificate.
    :::

## Configure Pomerium

1. Update the `config.yaml` file in `/etc/pomerium` with a new route for this upstream service:

    ```yaml

>     - from: https://openssl.pomerium.localhost.io

Reviewer 2: ```suggestion
    - from: https://openssl.localhost.pomerium.io
```

      to: https://localhost:44330
      tls_server_name: openssl.localhost
      policy:
        - allow:
            or:
              - email:
                  is: user@example.com
    ```

    Pomerium should automatically process changes when the configuration file is updated.

> 1. Visit <https://openssl.localhost.pomium.io> in your browser to confirm the route:

Reviewer 2: ```suggestion
1. Visit <https://openssl.localhost.pomerium.io> in your browser to confirm the route:
```

    ![OpenSSL Upstream without client certificate verification](./img/mtls/openssl-test1.png)

    Note the last line:

    > no client certificate available

    In the next section we'll adjust the OpenSSL server to require a client certificate and edit our Pomerium configuration to provide one.

## Install Pomerium Client Certificate

1. Stop the OpenSSL server process (**Ctrl+C**) and start a new one with the additional flag `-Verify 1`:

    ```bash
    openssl s_server -Verify 1 -key ./openssl.localhost-key.pem -cert ./openssl.localhost.pem -accept 44330 -www
    ```

1. When you refresh <https://openssl.localhost.pomium.io> in your browser, the connection will fail. Back in the terminal, the OpenSSL server should output errors containing:

    ```log
    SSL routines:tls_process_client_certificate:peer did not return a certificate
    ```

    This confirms that the upstream service is now requiring a trusted client certificate to allow connections.

1. Update the Pomerium route block to provide the client certificate:

    ```yaml
    - from: https://openssl.pomerium.localhost.io
      to: https://localhost:44330
      tls_server_name: openssl.localhost

>       tls_client_cert_file: /etc/pomerium/pomerium@localhost-client.pem
>       tls_client_key_file: /etc/pomerium/pomerium@localhost-client-key.pem

Reviewer 2: ```suggestion
      tls_client_cert_file: pomerium@localhost-client.pem
      tls_client_key_file: pomerium@localhost-client-key.pem
```

Author: Per discussion, this is required by the assumed environment, as with the other outstanding comment.

      policy:
        - allow:
            or:
              - email:
                  is: user@example.com

> 1. Now when you visit <https://openssl.localhost.pomium.io> you should see additional output under "Client certificate", confirming that the upstream service has read and validating Pomerium's client certificate. Your upstream service is now using mTLS for mutual authentication!

Reviewer: ```suggestion
1. Now when you visit <https://openssl.localhost.pomium.io> you should see additional output under "Client certificate", confirming that the upstream service has read and validates Pomerium's client certificate. Your upstream service is now using mTLS for mutual authentication!
```

Author: Good catch, but to match tense:
```suggestion
1. Now when you visit <https://openssl.localhost.pomium.io> you should see additional output under "Client certificate", confirming that the upstream service has read and validated Pomerium's client certificate. Your upstream service is now using mTLS for mutual authentication!
```
